In [1]:
import numpy as np
import pandas as pd
from sklearn import datasets, linear_model
from sklearn.preprocessing import scale
import statsmodels.api as sm
import statsmodels.formula.api as smf
from scipy import stats

In [2]:
%matplotlib notebook
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import axes3d # It is not explicitly used but is required for 3d projections 
import seaborn as sns

In [3]:
#Address
path = 'C:/Users/kiawo/OneDrive/Documents/ML_learning/An_Introduction_to_Statistical_Learning/Data/'
file = 'Advertising.csv'

In [4]:
df_raw = pd.read_csv(path+file, index_col=0) # for removing Unnamed:0

In [5]:
df_raw.head()

TV  radio  newspaper  sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

$$RSS = \sum_{i=1}^n (y_i - \hat y_i)^2 $$

## Ordinary Linear Regression

In [6]:
X_raw = df_raw['TV']
y = df_raw['sales']

In [7]:
#adding constant for having intercept in sm
X_with_const = sm.add_constant(X_raw)

C:\Users\kiawo\AppData\Roaming\Python\Python36\site-packages\numpy\core\fromnumeric.py:2495: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [9]:
X_with_const.head()

const     TV
1    1.0  230.1
2    1.0   44.5
3    1.0   17.2
4    1.0  151.5
5    1.0  180.8

In [10]:
est = sm.OLS(y, X_with_const).fit()
est.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Wed, 22 Jan 2020   Prob (F-statistic):           1.47e-42
Time:                        22:25:53   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.0326      0.458     15.360      0.000       6.130       7.935
TV             0.0475      0.003     17.668      0.000       0.042       0.053
==============================================================================
Omnibus:                        0.531   Durbin-Watson:                   1.935
Prob(Omnibus):                  0.767   Jarque-Bera (JB):                0.669
Skew:                          -0.089   Prob(JB):                        0.716
Kurtosis:                       2.779   Cond. No.                         338.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Explanation:

**DF Residual** – These are the degrees of freedom associated with the sources of variance.  The total variance has N-1 degrees of freedom.  The model degrees of freedom corresponds to the number of coefficients estimated minus 1.  Including the intercept, there are 2 coefficients, so the model has 2-1=1 degrees of freedom.  The Residual degrees of freedom is the DF total minus the DF model, 199 – 1 =198.

**R-squared** – R-Squared is the proportion of variance in the dependent variable (sale) which can be explained by the independent variables (TV).  This is an overall measure of the strength of association and does not reflect the extent to which any particular independent variable is associated with the dependent variable.

**Adj R-squared** – This is an adjustment of the R-squared that penalizes the addition of extraneous predictors to the model. Adjusted R-squared is computed using the below formula where p is the number of predictors.

$$ Adj R^2 = {1-(1-R^2){n-1 \over n-p-1}}$$


**F-statistic** – This is the  is the Mean Square Model divided by the Mean Square Residual, yielding F=312.1.

An F statistic is a value you get when you run an ANOVA test or a regression analysis to find out if the means between two populations are significantly different. It’s similar to a T statistic from a T-Test; A-T test will tell you if a single variable is statistically significant and an F test will tell you if a group of variables are jointly significant.

**Prob (F-statistic)** - This is the p-value associated with the above F-statistic. It is used in testing the null hypothesis that all of the model coefficients are 0.

If you have significant result, it means that your results likely did not happen by chance. If you don’t have statistically significant results, you throw your test data out (as it doesn’t show anything!); in other words, you can’t reject the null hypothesis.

**AIC** - The Akaike information criterion (AIC) is an estimator of the relative quality of statistical models for a given set of data. Given a collection of models for the data, AIC estimates the quality of each model, relative to each of the other models. Thus, AIC provides a means for model selection.

**BIC** - In statistics, the Bayesian information criterion (BIC) or Schwarz criterion (also SBC, SBIC) is a criterion for model selection among a finite set of models; the model with the lowest BIC is preferred. It is based, in part, on the likelihood function and it is closely related to the Akaike information criterion (AIC).

AIC and BIC hold the same interpretation in terms of model comparison. That is, the larger difference in either AIC or BIC indicates stronger evidence for one model over the other (the lower the better).

**t**  – These are the t-statistics used in testing whether a given coefficient is significantly different from zero.

**P>|t|** – This column shows the 2-tailed p-values used in testing the null hypothesis that the coefficient (parameter) is 0. Using an alpha of 0.05:

  - The coefficient for TV is significantly different from 0 because its p-value is 0.000, which is smaller than 0.05. 

  - The constant (const) is significantly different from 0 at the 0.05 alpha level.

**[0.025	0.975]** – These are the 95% confidence intervals for the coefficients.  The confidence intervals are related to the p-values such that the coefficient will not be statistically significant at alpha = .05 if the 95% confidence interval includes zero.  These confidence intervals can help you to put the estimate from the coefficient into perspective by seeing how much the value could vary.

___sourced from: https://stats.idre.ucla.edu/stata/output/regression-analysis-2/ ___

In [ ]:
sns.regplot(df_raw.TV, df_raw.sales, order=1, ci=None, scatter_kws={'color':'r', 's':9})
plt.xlim(-10,310)
plt.ylim(ymin=0);

### RSS

In [ ]:
# Regression coefficients (Ordinary Least Squares)
regr = linear_model.LinearRegression()

X = scale(df_raw.TV, with_mean=True, with_std=False).reshape(-1,1)
y = df_raw.sales

regr.fit(X,y)
print(regr.intercept_)
print(regr.coef_)

In [ ]:
reg = linear_model.Ridge (alpha = .5)
reg.fit(X,y)

In [ ]:
print(reg.intercept_)
print(reg.coef_)

In [ ]:
reg = linear_model.Lasso (alpha = .1)
reg.fit(X,y)

In [ ]:
print(reg.intercept_)
print(reg.coef_)

In [ ]:
# Create grid coordinates for plotting
B0 = np.linspace(regr.intercept_-2, regr.intercept_+2, 50)
B1 = np.linspace(regr.coef_-0.02, regr.coef_+0.02, 50)
xx, yy = np.meshgrid(B0, B1, indexing='xy')
Z = np.zeros((B0.size,B1.size))

# Calculate Z-values (RSS) based on grid of coefficients
for (i,j),v in np.ndenumerate(Z):
    Z[i,j] =((y - (xx[i,j]+X.ravel()*yy[i,j]))**2).sum()/1000

# Minimized RSS
min_RSS = r'$\beta_0$, $\beta_1$ for minimized RSS'
min_rss = np.sum((regr.intercept_+regr.coef_*X - y.values.reshape(-1,1))**2)/1000
min_rss

In [ ]:
fig = plt.figure(figsize=(15,6))
fig.suptitle('RSS - Regression coefficients', fontsize=20)

ax1 = fig.add_subplot(121)
ax2 = fig.add_subplot(122, projection='3d')

# Left plot
CS = ax1.contour(xx, yy, Z, cmap=plt.cm.Set1, levels=[2.15, 2.2, 2.3, 2.5, 3])
ax1.scatter(regr.intercept_, regr.coef_[0], c='r', label=min_RSS)
ax1.clabel(CS, inline=True, fontsize=10, fmt='%1.1f')

# Right plot
ax2.plot_surface(xx, yy, Z, rstride=3, cstride=3, alpha=0.3)
ax2.contour(xx, yy, Z, zdir='z', offset=Z.min(), cmap=plt.cm.Set1,
            alpha=0.4, levels=[2.15, 2.2, 2.3, 2.5, 3])
ax2.scatter3D(regr.intercept_, regr.coef_[0], min_rss, c='r', label=min_RSS)
ax2.set_zlabel('RSS')
ax2.set_zlim(Z.min(),Z.max())
ax2.set_ylim(0.02,0.07)

# settings common to both plots
for ax in fig.axes:
    ax.set_xlabel(r'$\beta_0$', fontsize=17)
    ax.set_ylabel(r'$\beta_1$', fontsize=17)
    ax.set_yticks([0.03,0.04,0.05,0.06])
    ax.legend()

## Multiple Linear Regression

In Ordinary Least Squares Regression with a single variable we described the relationship between the predictor and the response with a straight line. In the case of multiple regression we extend this idea by fitting a  p-dimensional hyperplane to our p predictors.

In [ ]:
X_raw = df_raw[['TV','radio','newspaper']]
y = df_raw['sales']

In [ ]:
#adding constant for having intercept in sm
X_with_const = sm.add_constant(X_raw)

In [ ]:
est = sm.OLS(y, X_with_const).fit()
est.summary()

### Correlation between variables

In [ ]:
df_raw.corr()

### Interaction

Interaction effects can be account for by including a new feature comprising the product of corresponding values from the interacting features:

$$\hat{y} = \hat{\beta_0} + \hat{\beta_1}x_1+ \hat{\beta_2}x_2 + \hat{\beta_3}x_1x_2 + \epsilon\,$$

where $x_1$ and $x_2$ are the values of the two features, respectively and $x_1x_2$ represents the interaction between the two. It can be useful to use scikit-learn’s **PolynomialFeatures** to creative interaction terms for all combination of features. 

In [ ]:
X_int = df_raw[['TV','radio','newspaper']]
y = df_raw['sales']

In [ ]:
X_int['TV_radio'] = X_int.TV*X_int.radio
X_int.head()

In [ ]:
#adding constant for having intercept in sm
X_with_const = sm.add_constant(X_int)

In [ ]:
est = sm.OLS(y, X_with_const).fit()
est.summary()

It means $(96.8 - 89.7)/(100 - 89.7) = 69\%$ of variablity of **sale** that remains after fitting the model has been explained by interaction term

### Polynomial

In [ ]:
#Address
#path = '/Users/ljn197/Machine_learning/Data/'
file = 'Auto.csv'

In [ ]:
df_raw = pd.read_csv(path+file)
df_raw.head()

In [ ]:
#replacing ? with NaN
df_raw = df_raw.replace('?', np.NaN)

In [ ]:
df_raw['horsepower_c'] = df_raw['horsepower'].astype(np.float)

In [ ]:
sns.lmplot(x = 'horsepower_c',y='mpg', data=df_raw, fit_reg=False)

Creating extra variable to accomodate polynomial:
$$mpg = \beta_0 + \beta_1\times horsepower+ \beta_2 \times horsepower^2 + \epsilon\,$$

In [ ]:
df_raw['horsepower_2'] = df_raw['horsepower_c'].pow(2)


In [ ]:
feed = df_raw[['horsepower_c','horsepower_2','mpg']].dropna()

In [ ]:
X_int = feed[['horsepower_c','horsepower_2']]
y = feed['mpg']

In [ ]:
#adding constant for having intercept in sm
X_with_const = sm.add_constant(X_int)

In [ ]:
est = sm.OLS(y, X_with_const).fit()
est.summary()

In [ ]:
y_pred = est.predict(X_with_const)

In [ ]:
sns.regplot(X_with_const['horsepower_c'], y_pred, fit_reg=False)

In [ ]:
fig, axs = plt.subplots(ncols=2)
sns.regplot(X_with_const['horsepower_c'], y_pred, fit_reg=False, ax=axs[0])
sns.regplot(x = 'horsepower_c',y='mpg', data=df_raw, fit_reg=False, ax=axs[1])

In [ ]:
fig, axs = plt.subplots()
sns.regplot(X_with_const['horsepower_c'], y_pred, fit_reg=False, ax=axs)
sns.regplot(x = 'horsepower_c',y='mpg', data=df_raw, fit_reg=False, ax=axs)

### Regression with Scikit-learn

In [ ]:
# Create linear regression object
regr = linear_model.LinearRegression()

# Train the model using the training sets
regr.fit(X_int, y)

# Make predictions using the testing set
y_pred_scikit = regr.predict(X_int) #scikit learn doesn't need the adding of constant

In [ ]:
type(y_pred_scikit)

In [ ]:
fig, axs = plt.subplots()
sns.regplot(X_int['horsepower_c'], y, fit_reg=False, ax=axs)
sns.regplot(x = X_int['horsepower_c'].values ,y=y_pred_scikit, fit_reg=True, order=2, scatter = True, ax=axs)

### comparing results

In [ ]:
df_comp = pd.DataFrame(data =[X_int.horsepower_c.values, y_pred, y_pred_scikit.tolist()],index=['horsepower','Statsmodel','Scikit'])
df_comp

In [ ]:
#Residual
est.fittedvalues

The most useful way to plot the residuals, though, is with your predicted values on the x-axis, and your residuals on the y-axis.

In [ ]:
sns.residplot(x = y_pred, y= est.resid_pearson, lowess=True, color="y")

In [ ]:
sns.regplot(y_pred_scikit, y_pred_scikit-y, color='b',label='Residual Plot')

In [ ]:
#Sum of squared (whitened) residuals.
est.ssr

In [ ]:
#Explained sum of squares
est.ess

In [ ]:
#Uncentered sum of squares. Sum of the squared values of the (whitened) endogenous response variable.
est.uncentered_tss

In [ ]:
#The total (weighted) sum of squares centered about the mean.
est.centered_tss

In [ ]:
#Total mean squared error. Defined as the uncentered total sum of squares divided by n the number of observations.
est.mse_total

In [ ]:
#Mean squared error the model. This is the explained sum of squares divided by the model degrees of freedom.
est.mse_model

In [ ]:
#Mean squared error of the residuals. The sum of squared residuals divided by the residual degrees of freedom.
est.mse_resid

In [ ]:
est.resid_pearson # Residuals, normalized to have unit variance.